File: src/main/java/com/telegram/codex/conversation/infrastructure/update/ProcessedUpdateRepository.java
package com.telegram.codex.conversation.infrastructure.update;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.telegram.codex.conversation.application.port.out.ProcessedUpdatePort;
import com.telegram.codex.conversation.application.reply.ReplyResult;
import com.telegram.codex.conversation.domain.update.ProcessedUpdateRecord;
import com.telegram.codex.conversation.infrastructure.persistence.ProcessedUpdateEntity;
import com.telegram.codex.conversation.infrastructure.persistence.ProcessedUpdateJpaRepository;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Repository;

import java.time.Duration;
import java.util.Optional;

@Repository
public class ProcessedUpdateRepository implements ProcessedUpdatePort {

    private static final long INFLIGHT_TIMEOUT_MS = Duration.ofMinutes(5).toMillis();

    private final ObjectMapper objectMapper;
    private final ProcessedUpdateJpaRepository repository;

    public ProcessedUpdateRepository(ObjectMapper objectMapper, ProcessedUpdateJpaRepository repository) {
        this.objectMapper = objectMapper;
        this.repository = repository;
    }

    @Override
    public Optional<ProcessedUpdateRecord> find(long updateId) {
        return repository.findById(updateId).map(this::toRecord);
    }

    @Transactional
    @Override
    public boolean beginProcessing(long updateId, String chatId, long messageId) {
        long now = System.currentTimeMillis();
        Optional<ProcessedUpdateEntity> existing = repository.findById(updateId);
        if (existing.isEmpty()) {
            ProcessedUpdateEntity entity = new ProcessedUpdateEntity();
            entity.setUpdateId(updateId);
            entity.setChatId(chatId);
            entity.setMessageId(messageId);
            entity.setProcessedAt(now);
            repository.save(entity);
            return true;
        }

        ProcessedUpdateEntity entity = existing.get();
        if (entity.getSentAt() != null) {
            return false;
        }
        if (entity.getReplyText() != null && entity.getConversationState() != null) {
            return false;
        }
        if (now - entity.getProcessedAt() < INFLIGHT_TIMEOUT_MS) {
            return false;
        }

        entity.setChatId(chatId);
        entity.setMessageId(messageId);
        entity.setProcessedAt(now);
        entity.setReplyText(null);
        entity.setConversationState(null);
        entity.setSuggestedReplies(null);
        entity.setSentAt(null);
        repository.save(entity);
        return true;
    }

    @Transactional
    @Override
    public void clearProcessing(long updateId) {
        repository.findById(updateId).ifPresent(entity -> {
            if (entity.getSentAt() == null && entity.getReplyText() == null && entity.getConversationState() == null) {
                repository.delete(entity);
            }
        });
    }

    @Transactional
    @Override
    public void markProcessed(long updateId, String chatId, long messageId) {
        ProcessedUpdateEntity entity = repository.findById(updateId).orElseGet(ProcessedUpdateEntity::new);
        long now = System.currentTimeMillis();
        entity.setUpdateId(updateId);
        entity.setChatId(chatId);
        entity.setMessageId(messageId);
        entity.setProcessedAt(now);
        entity.setSentAt(now);
        repository.save(entity);
    }

    @Transactional
    @Override
    public void savePendingReply(long updateId, String chatId, long messageId, ReplyResult result) {
        ProcessedUpdateEntity entity = repository.findById(updateId).orElseGet(ProcessedUpdateEntity::new);
        entity.setUpdateId(updateId);
        entity.setChatId(chatId);
        entity.setMessageId(messageId);
        entity.setProcessedAt(System.currentTimeMillis());
        entity.setReplyText(result.text());
        entity.setConversationState(result.conversationState());
        entity.setSuggestedReplies(writeSuggestedReplies(result.suggestedReplies()));
        entity.setSentAt(null);
        repository.save(entity);
    }

    @Override
    public long pruneSentBefore(long cutoff) {
        return repository.deleteBySentAtIsNotNullAndProcessedAtLessThan(cutoff);
    }

    private String writeSuggestedReplies(Object value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JsonProcessingException error) {
            throw new IllegalStateException("Failed to persist suggested replies", error);
        }
    }

    private ProcessedUpdateRecord toRecord(ProcessedUpdateEntity entity) {
        return new ProcessedUpdateRecord(
            entity.getUpdateId(),
            entity.getChatId(),
            entity.getMessageId(),
            entity.getProcessedAt(),
            entity.getReplyText(),
            entity.getConversationState(),
            entity.getSuggestedReplies(),
            entity.getSentAt()
        );
    }
}

File: src/main/kotlin/com/telegram/codex/conversation/infrastructure/ProcessedUpdateRepository.kt
package com.telegram.codex.conversation.infrastructure

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.telegram.codex.conversation.application.ReplyResult
import com.telegram.codex.conversation.domain.ProcessedUpdateRecord
import jakarta.transaction.Transactional
import org.springframework.stereotype.Repository
import java.time.Duration
import java.util.*

@Repository
class ProcessedUpdateRepository(
    private val objectMapper: ObjectMapper,
    private val repository: ProcessedUpdateJpaRepository,
) {
    fun find(updateId: Long): Optional<ProcessedUpdateRecord> = repository.findById(updateId).map(::toRecord)

    @Transactional
    fun beginProcessing(updateId: Long, chatId: String, messageId: Long): Boolean {
        val now = System.currentTimeMillis()
        val existing = repository.findById(updateId)
        if (existing.isEmpty) {
            val entity = ProcessedUpdateEntity()
            entity.updateId = updateId
            entity.chatId = chatId
            entity.messageId = messageId
            entity.processedAt = now
            repository.save(entity)
            return true
        }

        val entity = existing.get()
        if (entity.sentAt != null) {
            return false
        }
        if (entity.replyText != null && entity.conversationState != null) {
            return false
        }
        if (now - entity.processedAt < INFLIGHT_TIMEOUT_MS) {
            return false
        }

        entity.chatId = chatId
        entity.messageId = messageId
        entity.processedAt = now
        entity.replyText = null
        entity.conversationState = null
        entity.suggestedReplies = null
        entity.sentAt = null
        repository.save(entity)
        return true
    }

    @Transactional
    fun clearProcessing(updateId: Long) {
        repository.findById(updateId).ifPresent { entity ->
            if (entity.sentAt == null && entity.replyText == null && entity.conversationState == null) {
                repository.delete(entity)
            }
        }
    }

    @Transactional
    fun markProcessed(updateId: Long, chatId: String, messageId: Long) {
        val entity = repository.findById(updateId).orElseGet(::ProcessedUpdateEntity)
        val now = System.currentTimeMillis()
        entity.updateId = updateId
        entity.chatId = chatId
        entity.messageId = messageId
        entity.processedAt = now
        entity.sentAt = now
        repository.save(entity)
    }

    @Transactional
    fun savePendingReply(updateId: Long, chatId: String, messageId: Long, result: ReplyResult) {
        val entity = repository.findById(updateId).orElseGet(::ProcessedUpdateEntity)
        entity.updateId = updateId
        entity.chatId = chatId
        entity.messageId = messageId
        entity.processedAt = System.currentTimeMillis()
        entity.replyText = result.text
        entity.conversationState = result.conversationState
        entity.suggestedReplies = writeSuggestedReplies(result.suggestedReplies)
        entity.sentAt = null
        repository.save(entity)
    }

    fun pruneSentBefore(cutoff: Long): Long = repository.deleteBySentAtIsNotNullAndProcessedAtLessThan(cutoff)

    private fun writeSuggestedReplies(value: Any?): String =
        try {
            objectMapper.writeValueAsString(value)
        } catch (error: JsonProcessingException) {
            throw IllegalStateException("Failed to persist suggested replies", error)
        }

    private fun toRecord(entity: ProcessedUpdateEntity): ProcessedUpdateRecord =
        ProcessedUpdateRecord(
            entity.updateId ?: 0L,
            entity.chatId,
            entity.messageId ?: 0L,
            entity.processedAt,
            entity.replyText,
            entity.conversationState,
            entity.suggestedReplies,
            entity.sentAt,
        )

    companion object {
        private val INFLIGHT_TIMEOUT_MS = Duration.ofMinutes(5).toMillis()
    }
}
